### PR TITLE
[homekit] add support for complex accessories

### PR DESCRIPTION
**Pull request by yfre**
HomeKit accessory can be combination of multiple accessories, e.g. fan with light, sensor with battery, irrigation with multiple valves
iOS home app would show such complex accessory as one tile on the home screen that opens a view with multiple controls. 

this PR adds support for such complex accessories. 

in openHAB complex accessory would be modelled by adding additional HomeKit tag to the group metadata, e.g. 

```xtend
Group           FanWithLight        "Fan with Light"                           {homekit = "Fan,Light"}
Switch          FanActiveStatus     "Fan Active Status"     (FanWithLight)     {homekit = "Fan.ActiveStatus"}
Number          FanRotationSpeed    "Fan Rotation Speed"    (FanWithLight)     {homekit = "Fan.RotationSpeed"}
Switch          Light               "Light"                 (FanWithLight)     {homekit = "Lighting.OnState"}
```

it is backward compatible. all existing HomeKit configurations have only one tag at group level and should work as before.

### CODE REVIEW

**Comment by yfre** [comment]
@andylintner @lolodomo 
could please review this PR. 
the UI part for this got merged already https://github.com/openhab/openhab-webui/pull/1317

im afraid, UI configurations that are not supported by the binding could confuse users.

**Review by andylintner — APPROVED** [review]
LGTM

File: bundles/org.openhab.io.homekit/README.md [hunk 1]
```diff
@@ -457,7 +457,73 @@ or using UI
 
 ![sensor_ui_config.png](doc/sensor_ui_config.png)
 
+### Complex accessory
 
+Multiple HomeKit accessories can be combined to one accessory in order to group several functions provided by one or multiple physical devices. 
+
+For example, ceiling fans often include lighting functionality. Such fans can be modelled as:
```
Comment (lolodomo) [inline bundles/org.openhab.io.homekit/README.md:464]: I think it is "modeled" with only one l

File: bundles/org.openhab.io.homekit/README.md [hunk 2]
```diff
@@ -457,7 +457,73 @@ or using UI
...
+Switch          Light               "Light"                 (FanWithLight)     {homekit = "Lighting.OnState"}
+```
+
+or in mainUI
+![ui_fan_with_light_group_view.png](doc/ui_fan_with_light_group_view.png)
+![ui_fan_with_light_group_code.png](doc/ui_fan_with_light_group_code.png)
+![ui_fan_with_light_group_config.png](doc/ui_fan_with_light_group_config.png)
+
+
+iOS home app use by default the type of the first accessory on the list for the tile on home screen.
```
Comment (lolodomo) [inline bundles/org.openhab.io.homekit/README.md:501]: ... uses ...

File: bundles/org.openhab.io.homekit/src/main/java/org/openhab/io/homekit/internal/HomekitChangeListener.java [hunk 3]
```diff
@@ -273,19 +281,66 @@ public int getConfigurationRevision() {
...
+     * - it has HomeKit accessory types defined using HomeKit accessory metadata
+     * - AND is part of groups with HomeKit metadata, but all groups have baseItem
+     * e.g.
+     * Group:Switch:OR(ON,OFF) gLight "Light Group " {homekit="Lighting"}
+     * Switch light "Light" (gLight) {homekit="Lighting.OnState"}
+     *
+     *
      * In contrast, items which are part of groups without BaseItem are additional HomeKit characteristics of the
-     * accessory defined by that group and dont need to be created as RootAccessory here.
+     * accessory defined by that group and dont need to be created as accessory here.
```
Comment (lolodomo) [inline bundles/org.openhab.io.homekit/src/main/java/org/openhab/io/homekit/internal/HomekitChangeListener.java:336]: ... don't ...

**Review by lolodomo — CHANGES_REQUESTED** [review]
Only few typo

**Comment by yfre** [comment]
> Only few typo

thank you [lolodomo](https://github.com/lolodomo) for the review.

**Comment by JacekKac** [comment]
Hi @yfre can you post an example of Valve grouping which are already in a group ?

```

Group                   gLawn1          "Zraszacz 1"                                            {homekit = "Valve" [homekitValveType="Irrigation", homekitTimer="true"]}
Switch                  Lawn1           "Valve active"                          (gLawn1)        {homekit = "Valve.ActiveStatus, Valve.InUseStatus"}
Number                  Lawn1_duration  "Valve duration"                        (gLawn1)        {homekit = "Valve.Duration" [homekitDefaultDuration = 300]}

Group                   gLawn2          "Zraszacz 2"                                            {homekit = "Valve" [homekitValveType="Irrigation", homekitTimer="true"]}
Switch                  Lawn2           "Valve active"                          (gLawn2)        {homekit = "Valve.ActiveStatus, Valve.InUseStatus"}
Number                  Lawn2_duration  "Valve duration"                        (gLawn2)        {homekit = "Valve.Duration" [homekitDefaultDuration = 300]}


```

**Comment by yfre** [comment]
> Hi @yfre can you post an example of Valve grouping which are already in a group ?
> 
> ```
> 
> Group                   gLawn1          "Zraszacz 1"                                            {homekit = "Valve" [homekitValveType="Irrigation", homekitTimer="true"]}
> Switch                  Lawn1           "Valve active"                          (gLawn1)        {homekit = "Valve.ActiveStatus, Valve.InUseStatus"}
> Number                  Lawn1_duration  "Valve duration"                        (gLawn1)        {homekit = "Valve.Duration" [homekitDefaultDuration = 300]}
> 
> Group                   gLawn2          "Zraszacz 2"                                            {homekit = "Valve" [homekitValveType="Irrigation", homekitTimer="true"]}
> Switch                  Lawn2           "Valve active"                          (gLawn2)        {homekit = "Valve.ActiveStatus, Valve.InUseStatus"}
> Number                  Lawn2_duration  "Valve duration"                        (gLawn2)        {homekit = "Valve.Duration" [homekitDefaultDuration = 300]}
> ```
so, if you want to get something like this then this is currently not possible. this requires one more accessory type called "IrrigationSystem" which is not supported by homekit binding yet.  i have it on my todo list. 

image

**Comment by JacekKac** [comment]
thank you, 
i think ( i have only photo of such configuration left) it was build on homebridge without irrigation item support (no settings button at the bottom). Let's try on other items - is it possible to configure 2 homekit = "Lightning" items Lamp 1 and Lamp 2 under one grouped accessories GUI allows to Select Multiple accessories (many types ) but only one type occuring once). If i put "homekit="Lighting, Lighting" in text config GUI addoptes and allows to define 2 separate Lighting groups but it's not working in Home app

**Comment by yfre** [comment]
> thank you, i think ( i have only photo of such configuration left) it was build on homebridge without irrigation item support (no settings button at the bottom). Let's try on other items - is it possible to configure 2 homekit = "Lightning" items Lamp 1 and Lamp 2 under one grouped accessories GUI allows to Select Multiple accessories (many types ) but only one type occuring once). If i put "homekit="Lighting, Lighting" in text config GUI addoptes and allows to define 2 separate Lighting groups but it's not working in Home app

multiple accessories of the same type is currently not supported. here we would need to have support group of groups as discussed here

https://community.openhab.org/t/does-openhab-support-grouping-homekit-accessories-in-different-types/132367/4?u=yfre

something like this

```
group combinedLights {homekit="Lighting, Lighting"}

group light1 (combinedLights) {homekit="Lighting"}
switch light1_switch (light1)

group light2  (combinedLights) {homekit="Lighting"}
switch light2_switch (light2)
```

it still looks complex to me but i dont have any better ideas for complex items of the same type

**Pull request merged**